### ISSUE TITLE
[RFC] Reduce complexity of Rector\Core\Contract\Processor\FileProcessorInterface::process

### ISSUE THREAD

**Issue by yguedidi** [rectorphp/rector#7937]
# TL;DR

Basically what is proposed in https://github.com/rectorphp/rector-src/pull/3739 but I feel the discussion is not concluded 😅
Or it's only me not being satisfied by the "[changing too much of architecture](https://github.com/rectorphp/rector-src/pull/3739#issuecomment-1537168154)" counter argument compared to benefits...
Sorry for that.. I know I link to understand things
Feel free to tell me to stop in case it's a definitive decision, it's fine 🙂 

# Context

This `Rector\Core\Contract\Processor\FileProcessorInterface::process` function is about processing a single `File`.
It has 2 builtin implementations: `PhpFileProcessor` and `NonPhpFileProcessor`.

Current API is:
```php
    /**
     * @return array{system_errors: SystemError[], file_diffs: FileDiff[]}
     */
    public function process(File $file, Configuration $configuration): array;
```
It returns an array of system errors and file diffs, in plural.

Looking at implementations, the current possible returns are:
- an empty array in both keys
- a single element in file diffs key
- a single element in system errors key

To be noted that only `PhpFileProcessor` can return system errors currently.

Simplified it returns either a `FileDiff`, either a `SystemError`, either nothing.
At this stage we could consider a better API a return type of `FileDiff|SystemError|null`.

The de-facto way of managing user-land errors in PHP is using exceptions.

The only type of system error that can be returned by `PhpFileProcessor` are parsing errors:

https://github.com/rectorphp/rector-src/blob/main/src/Application/FileProcessor/PhpFileProcessor.php#L62

We can throw a `ParsingException` in such case, to let error management be done in the right place needing it, the 2 entry points that are:
- `Rector\Parallel\WorkerRunner` for parallel processing
- `Rector\Core\Application\ApplicationFileProcessor` for non-parallel processing

We end up with this proposed API:
```php
    /**
     * @throws ParsingException
     */
    public function process(File $file, Configuration $configuration): ?FileDiff;
```

# Benefits

- cleaner API, following SRP, expressing better what the interface does
- reduced complexity of the interface
- reduced complexity of implementations
- error management following best practices, being centralized in a high level place, the entry points

# Potential issues

It's a contract, if there is any other existing implementation it will need to be updated.
But we are at 0.x version so BC breaks are allowed.

# Alternative

We can start at least by removing the plural confusion, without the better error throwing, by changing the return type to `FileDiff|SystemError|null`.
But this alternative feels "incomplete" to me, even tho the API is slightly better this way.

**Comment by samsonasik** [comment]
`FileProcessorInterface` is an interface, that should not be changed as possible to avoid direct break the consumer on rector extension.

If such change is needed, old hahaviour should be marked as `@deprecated` and call new method.

**Comment by TomasVotruba** [comment]
Hi, I want to keep this interface as it is, 
because it works well and we know how to work with it.

**Comment by yguedidi** [comment]
@samsonasik thanks for the fast answer!
I'm aware of that as mentioned in the Potential issues section.

To me here all the benefits are worth a fast update on those extensions using the interface.

Also, it depends on how those extensions currently require Rector, if they require a tagged version of the dev version.
If the require a tagged version (that they should do IMO), it's fine then, an update to them will bump to the higher version of Rector including it.

**Comment by yguedidi** [comment]
@TomasVotruba OK, maintainer's decision.
Did my best to argument the proposal based on your feedback regarding architectural changes, that they should reduce complexity etc...
Your project, your choices.

A bit sad that this isn't accept, but taking it positively it was a good exercice to find arguments for the proposal.
More RFCs to come!
